text/0042-partial-fill-rules.md

```diff
@@ -0,0 +1,296 @@
+InputRules: RevealedTxOut partial_fill_change *new*
+InputRules: uint64 min_partial_fill_value *new*
+
+RevealedTxOut: TxOut tx_out
+RevealedTxOut: bytes amount_shared_secret
+```
+
+Recall that the `InputRules` are created (and signed) by the **originator**, and this TxIn is added to a transaction
+by the **counterparty**, who is also obligated to add TxOut's to the TxPrefix which match to the partial fill change
+and partial fill outputs appropriately. (This is actually done under the hood by the transaction builder.)
+
+The consensus enclave must enforce any input rules, including these new partial fill rules. This is done in the following way.
+
+1. If `partial_fill_change` is not present, then no other partial fill rules may be used, and any other partial fill fields must be defaulted.
+1. If `partial_fill_change` is present, then decrypting the amount and token id using `amount_shared_secret` must succeed. The partial fill change
+   may not have a 0 value (to help clients avoid division by zero).
+1. If `partial_fill_change` is present, then a corresponding TxOut called the `fractional_change_output` must exist in the `TxPrefix.outputs`.
+   This must match the `partial_fill_change` in every field except possibly the `masked_amount`.
+1. The `fractional_change_output` must decrypt successfully using the `amount_shared_secret` of the `partial_fill_change`, and
+   it's amount must have the same token id as the `partial_fill_change`, and its value must be less or equal.
+1. The fill fraction is inferred at this point.
+   * The numerator is `partial_fill_change.value - fractional_change_output.value`.
+   * The denominator is `partial_fill_change.value`.
+1. The minimum fill rule is enforced: `partial_fill_change.value - fractional_change_output.value >= min_partial_fill_value`.
+   If this inequality does not hold, then the transaction is rejected for reasons of not meeting the minimum partial fill prescribed
+   by the origiantor.
+1. For each `partial_fill_output`:
+   * It must decrypt succesfully using the recorded `amount_shared_secret`.
+   * There must be a corresponding `fractional_output` among the `TxPrefix.outputs` which matches it in every field except possibly
+     the `masked_amount`.
+   * The fractional output must decrypt successfully using the `partial_fill_output`'s `amount_shared_secret`, and must match the
+     token id of the `partial_fill_output`.
+   * It must be the case that `fractional_output.value >= n/d * partial_fill_output.value` where `n` and `d` are the fill fraction
+     numerator and denominator. This equation must be validated by clearing denominators and checking an inequality of `u128`'s, to
+     avoid numerical issues.
+     
+     ```
+     fractional_output.value * fill_fraction_denominator >= fill_fraction_numerator * partial_fill_output.value
+     ```
+     
+     If this inequality does not hold, then the transaction must be rejected for reasons of not respecting the fill fraction.
```

Comment:
:wrench: Even in the code I think this was a bit hard to follow.
I'm not sure it's much better, especially since it's already nested 2 levels.  Also it may be a misuse of **greifing**
```suggestion
   * It must be the case that:
      
     $$ {\texttt{partial_fill_change.value} - \texttt{fractional_change_output.value} \over \texttt{partial_fill_change.value}} <= {\texttt{fractional_output.value} \over \texttt{partial_fill_output.value}} $$
     
     **Note** Ideally we want these to be equal, but due to precision we favor the originator to prevent the counterparty from choosing values that allows griefing the originator.
     
     In the implementation the division is avoided, to further reduce precision loss:
     
     $$ ({\texttt{partial_fill_change.value} - \texttt{fractional_change_output.value}) \times \texttt{partial_fill_output.value}} <= {\texttt{fractional_output.value} \times \texttt{partial_fill_change.value}} $$
     
     If this inequality does not hold, then the transaction must be rejected for reasons of not respecting the fill fraction.
```

Reply:
I largely took this change, except for the latex part